net/server.go

```diff
@@ -447,3 +461,136 @@ func (s *server) SendPubSubMessage(
 	}
 	return t.Publish(ctx, data)
 }
+
+// hasAccess checks if the requesting peer has access to the given ci.
+//
+// This is used as a filter in bitswap to determine if we should send the block to the requesting peer.
+func (s *server) hasAccess(p libpeer.ID, c cid.Cid) bool {
+	if !s.peer.acp.HasValue() {
+		return true
+	}
+
+	rawblock, err := s.peer.blockstore.Get(s.peer.ctx, c)
+	if err != nil {
+		log.ErrorE("Failed to get block", err)
+		return false
+	}
+	block, err := coreblock.GetFromBytes(rawblock.RawData())
+	if err != nil {
+		log.ErrorE("Failed to get doc from block", err)
+		return false
+	}
+
+	cols, err := s.peer.db.GetCollections(
+		s.peer.ctx,
+		client.CollectionFetchOptions{
+			SchemaVersionID: immutable.Some(block.Delta.GetSchemaVersionID()),
+		},
+	)
+	if err != nil {
+		log.ErrorE("Failed to get collections", err)
+		return false
+	}
+	if len(cols) == 0 {
+		log.Info("No collections found", corelog.Any("Schema Version ID", block.Delta.GetSchemaVersionID()))
+		return false
+	}
+
+	// If the requesting peer is in the replicators list for that collection, then they have access.
+	if peerList, ok := s.replicators[cols[0].SchemaRoot()]; ok {
+		_, exists := peerList[p]
+		if exists {
+			return true
+		}
+	}
+
+	identFunc := func() immutable.Option[identity.Identity] {
+		s.piMux.RLock()
+		ident, ok := s.peerIdentities[p]
+		s.piMux.RUnlock()
+		if !ok {
+			resp, err := s.getIdentity(s.peer.ctx, p)
+			if err != nil {
+				log.ErrorE("Failed to get identity", err)
+				return immutable.None[identity.Identity]()
+			}
+			ident, err = identity.FromToken(resp.IdentityToken)
+			if err != nil {
+				log.ErrorE("Failed to parse identity token", err)
+				return immutable.None[identity.Identity]()
+			}
+			err = identity.VerifyAuthToken(ident, s.peer.PeerID().String())
+			if err != nil {
+				log.ErrorE("Failed to verify auth token", err)
+				return immutable.None[identity.Identity]()
+			}
+			s.piMux.Lock()
+			s.peerIdentities[p] = ident
+			s.piMux.Unlock()
+		}
+		return immutable.Some(ident)
+	}
+
+	peerHasAccess, err := permission.CheckDocAccessWithIdentityFunc(
+		s.peer.ctx,
+		identFunc,
+		s.peer.acp.Value(),
+		cols[0], // For now we assume there is only one collection.
+		acp.ReadPermission,
+		string(block.Delta.GetDocID()),
+	)
+	if err != nil {
+		log.ErrorE("Failed to check access", err)
+		return false
+	}
+
+	return peerHasAccess
+}
+
+// trySelfHasAccess checks if the local node has access to the given block.
+//
+// This is a best-effort check and returns true unless we explicitly find that we don't have access.
```

Comment:
todo: Please document why we are even bothering if a 'best effort' is correct enough.

If the reason is that a proper check will be carried out somewhere else (e.g. another node), please also document why false-negatives are either impossible, or are tolerable, as it looks like we could refuse to process a block that should actually be processed.

Reply:
> as it looks like we could refuse to process a block that should actually be processed.

I think you probably meant the opposite?

This method is for **self** check of access so it will only block processing if it can confirm no granted access. In any other case it will continue with the sync process.

Reply:
> I think you probably meant the opposite?

Nope.

This function looks like it can falsely claim that we dont have access (e.g. local node doesn't have access (yet), but the other remote node does), in which case the block will not be processed.

Reply:
> This function looks like it can falsely claim that we dont have access (e.g. local node doesn't have access (yet), but the other remote node does), in which case the block will not be processed.

`Doesn't have access yet` is a correct claim that we don't have access. It doesn't matter what the remote node has access to. Trying to process a block that we know we don't have access will just result in the remote node not giving us access anyways.

Reply:
Ah okay - can you add that to the func docs, along with a bit detailing as to why a 'best-effort' is relevant/okay here?

Reply:
It is documented already no?

Reply:
Ah yes it has been since I opened this thread, thanks!  LGTM :)